File: bookdata/tracking.py
"""
Code for supporting import data tracking and relationships.
"""

import hashlib
import logging
from io import StringIO
from pathlib import Path

from . import db

_log = logging.getLogger(__name__)


def hash_and_record_file(cur, path, stage=None):
    """
    Compute the checksum of a file and record it in the database.
    """
    h = hashlib.md5()
    with open(path, 'rb') as f:
        data = f.read(8192 * 4)
        while data:
            h.update(data)
            data = f.read(8192 * 4)
    hash = h.hexdigest()
    path = Path(path).as_posix()
    _log.info('recording file %s with hash %s', path, hash)
    record_file(cur, path, hash, stage)
    return hash


def begin_stage(cur, stage):
    """
    Record that a stage is beginning.
    """
    if hasattr(cur, 'cursor'):
        # this is a connection
        with cur, cur.cursor() as c:
            return begin_stage(c, stage)
    _log.info('starting or resetting stage %s', stage)
    cur.execute('''
        INSERT INTO stage_status (stage_name)
        VALUES (%s)
        ON CONFLICT (stage_name)
        DO UPDATE SET started_at = now(), finished_at = NULL, stage_key = NULL
    ''', [stage])
    cur.execute('DELETE FROM stage_file WHERE stage_name = %s', [stage])
    cur.execute('DELETE FROM stage_dep WHERE stage_name = %s', [stage])
    cur.execute('DELETE FROM stage_table WHERE stage_name = %s', [stage])


def record_dep(cur, stage, dep):
    """
    Record a dependency for a stage.
    """
    if hasattr(cur, 'cursor'):
        # this is a connection
        with cur, cur.cursor() as c:
            return record_dep(c, stage, dep)

    _log.info('recording dep %s -> %s', stage, dep);
    cur.execute('''
        INSERT INTO stage_dep (stage_name, dep_name, dep_key)
        SELECT %s, stage_name, stage_key
        FROM stage_status WHERE stage_name = %s
        RETURNING dep_name, dep_key
    ''', [stage, dep])
    return cur.fetchall()


def record_tbl(cur, stage, ns, tbl):
    """
    Record a table associated with a stage.
    """
    if hasattr(cur, 'cursor'):
        # this is a connection
        with cur, cur.cursor() as c:
            return record_tbl(c, stage, ns, tbl)

    _log.info('recording table %s -> %s.%s', stage, ns, tbl);
    cur.execute('''
        INSERT INTO stage_table (stage_name, st_ns, st_name)
        VALUES (%s, %s, %s)
    ''', [stage, ns, tbl])
    cur.execute('''
        SELECT oid, kind FROM stage_table_oids WHERE stage_name = %s AND st_ns = %s AND st_name = %s
    ''', [stage, ns, tbl])
    return cur.fetchone()


def record_file(cur, file, hash, stage=None):
    """
    Record a file and optionally associate it with a stage.
    """
    if hasattr(cur, 'cursor'):
        # this is a connection
        with cur, cur.cursor() as c:
            return record_file(c, stage)
    _log.info('recording checksum %s for file %s', hash, file)
    cur.execute("""
        INSERT INTO source_file (filename, checksum)
        VALUES (%(file)s, %(hash)s)
        ON CONFLICT (filename)
        DO UPDATE SET checksum = %(hash)s, reg_time = NOW()
        """, {'file': file, 'hash': hash})
    if stage is not None:
        cur.execute("INSERT INTO stage_file (stage_name, filename) VALUES (%s, %s)", [stage, file])


def end_stage(cur, stage, key=None):
    """
    Record that an import stage has finished.

    Args:
        cur(psycopg2.connection or psycopg2.cursor): the database connection to use.
        stage(string): the name of the stage.
        key(string or None): the key (checksum or other key) to record.
    """
    if hasattr(cur, 'cursor'):
        # this is a connection
        with cur, cur.cursor() as c:
            return end_stage(c, stage, key)
    _log.info('finishing stage %s', stage)
    cur.execute('''
        UPDATE stage_status
        SET finished_at = NOW(), stage_key = coalesce(%(key)s, stage_key)
        WHERE stage_name = %(stage)s
    ''', {'stage': stage, 'key': key})


def stage_status(stage, file=None, *, timestamps=False):
    if file is None:
        sf = StringIO()
    else:
        sf = file

    with db.connect() as dbc:
        # initialize database, in case nothing has been run
        with dbc, dbc.cursor() as cur:
            cur.execute(db.meta_schema)

        # get the status
        with dbc, dbc.cursor() as cur:
            cur.execute('''
                SELECT started_at, finished_at, stage_key FROM stage_status WHERE stage_name = %s
            ''', [stage])
            row = cur.fetchone()
            if not row:
                _log.error('stage %s not run', stage)
                sys.exit(2)

            start, end, key = row

            _log.info('stage %s finished at %s', stage, end)
            print('STAGE', stage, file=sf)
            if timestamps:
                print('START', start, file=sf)

            cur.execute('''
                SELECT dep_name, dep_key
                FROM stage_dep
                WHERE stage_name = %s
                ORDER BY dep_name
            ''', [stage])
            for dn, dk in cur:
                print('DEP', dn, dk, file=sf)

            cur.execute('''
                SELECT filename, COALESCE(link.checksum, src.checksum)
                FROM source_file src
                JOIN stage_file link USING (filename)
                WHERE stage_name = %s
                ORDER BY filename
            ''', [stage])
            for fn, fh in cur:
                print('SOURCE', fn, fh, file=sf)

            cur.execute('''
                SELECT st_ns, st_name, oid, kind
                FROM stage_table_oids
                WHERE stage_name = %s
                ORDER BY st_ns, st_name
            ''', [stage])
            for ns, tbl, oid, kind in cur:
                print(f'TABLE {ns}.{tbl} OID {oid} KIND {kind}', file=sf)

            if timestamps:
                print('FINISH', end, file=sf)
            if key:
                print('KEY', key, file=sf)

    if file is None:
        return sf.getvalue()

File: src/tracking.rs
use std::io;
use std::io::Write;
use std::path::{Path, PathBuf};
use std::fs::OpenOptions;

use anyhow::Result;
use sha1::Sha1;
use structopt::StructOpt;
use postgres::Connection;

use log::*;

use crate::io::HashRead;

/// Options controlling the import stage
#[derive(StructOpt, Debug, Clone)]
pub struct StageOpts {
  /// Stage name
  #[structopt(long="stage", short="s")]
  stage: Option<String>,

  /// Stage dependencies
  #[structopt(long="stage-dep", short="D")]
  deps: Vec<String>,

  /// Transcript file
  #[structopt(long="transcript", short="T")]
  transcript: Option<PathBuf>,
}

/// An import stage.  Writing to the stage writes to its transcript file.
pub struct Stage<'o, 'c> {
  options: &'o StageOpts,
  cxn: Option<&'c Connection>,
  transcript: Box<dyn io::Write>
}

/// A source file for a stage
pub struct StageSource<'s> {
  stage: &'s Stage<'s, 's>,
  path: String,
  hash: Sha1
}

impl StageOpts {
  /// Start the stage
  pub fn begin_stage<'o, 'c>(&'o self, cxn: &'c Connection) -> Result<Stage<'o, 'c>> {
    match self.stage {
      Some (ref s) => {
        info!("beginning stage {}", s);
        cxn.execute("INSERT INTO stage_status (stage_name)
                     VALUES ($1)
                     ON CONFLICT (stage_name)
                     DO UPDATE SET started_at = now(), finished_at = NULL, stage_key = NULL",
                    &[s])?;
        cxn.execute("DELETE FROM stage_file WHERE stage_name = $1", &[s])?;
        cxn.execute("DELETE FROM stage_dep WHERE stage_name = $1", &[s])?;
        for d in &self.deps {
          cxn.execute("INSERT INTO stage_dep (stage_name, dep_name, dep_key)
                       SELECT $1, stage_name, stage_key
                       FROM stage_status WHERE stage_name = $2", &[s, &d])?;
        }
      },
      None => {
        warn!("no stage specified");
      }
    };
    let w: Box<dyn io::Write> = match self.transcript {
      Some(ref p) => {
        Box::new(OpenOptions::new().write(true).create(true).truncate(true).open(p)?)
      },
      None => Box::new(io::stdout())
    };
    Ok(Stage {
      options: self,
      cxn: Some(cxn),
      transcript: w
    })
  }

  /// Create a no-op stage.
  pub fn empty<'o>(&'o self) -> Stage<'o, 'o> {
    Stage {
      options: self,
      cxn: None,
      transcript: Box::new(io::stderr())
    }
  }
}

impl <'o, 'c> Write for Stage<'o, 'c> {
  fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
    self.transcript.write(buf)
  }

  fn flush(&mut self) -> io::Result<()> {
    self.transcript.flush()
  }

  fn write_fmt(&mut self, fmt: std::fmt::Arguments) -> io::Result<()> {
    self.transcript.write_fmt(fmt)
  }
}

impl <'o,'c> Stage<'o,'c> {
  /// End the stage
  pub fn end(self, key: &Option<String>) -> Result<()> {
    match self.options.stage {
      Some (ref s) => {
        info!("finishing stage {}", s);
        self.db_action(|db| {
          Ok(db.execute("UPDATE stage_status
                         SET finished_at = NOW(), stage_key = $2
                         WHERE stage_name = $1",
                        &[s, &key])?)
        })?;
      },
      None => ()
    };
    Ok(())
  }

  fn db_action<F, R, Rt>(&self, func: F) -> Result<Option<R>> where F: FnOnce(&Connection) -> Rt, Rt: Into<Result<R>> {
    match self.cxn {
      Some(ref c) => Ok(Some(func(c).into()?)),
      None => Ok(None)
    }
  }

  /// Set up to record a file with its reader, to both source and transcript
  pub fn source_file<'s, P: AsRef<Path>>(&'s self, path: P) -> StageSource<'s> {
    let path: &Path = path.as_ref();
    StageSource {
      stage: self,
      path: path.to_string_lossy().to_string(),
      hash: Sha1::new()
    }
  }
}

impl <'s> StageSource<'s> {
  /// Wrap a reader to compute this file's hash
  pub fn wrap_read<'a, R: io::Read>(&'a mut self, read: R) -> HashRead<'a, R> {
    HashRead::create(read, &mut self.hash)
  }

  /// Record the accumulated file hash (and return it)
  pub fn record(self) -> Result<String> {
    let hash = self.hash.hexdigest();
    self.record_hash(&hash)?;
    Ok(hash)
  }

  fn record_hash(&self, hash: &str) -> Result<()> {
    info!("recording checksum {} for file {}", hash, &self.path);
    self.stage.db_action(|db| {
      db.execute("INSERT INTO source_file (filename, checksum)
                  VALUES ($1, $2)
                  ON CONFLICT (filename)
                  DO UPDATE SET checksum = $2, reg_time = NOW()",
                 &[&self.path, &hash])?;
      match self.stage.options.stage {
        Some (ref s) => {
          debug!("attaching to stage {}", s);
          db.execute("INSERT INTO stage_file (stage_name, filename)
                      VALUES ($1, $2)",
                     &[s, &self.path])?
        },
        None => 0
      };
      Ok(())
    })?;
    Ok(())
  }
}
